src/main/scala/rise/core/uniqueNames.scala

```diff
@@ -1,51 +1,29 @@
 package rise.core
 
+import rise.core.Traverse.{Pure, PureTraversal}
 import rise.core.types._
 
 object uniqueNames {
   private case class CountingVisitor(
-    values: Map[Identifier, Int],
-    others: Map[Kind.Identifier, Int]
-  ) extends traversal.Visitor {
-    override def visitExpr(e: Expr): traversal.Result[Expr] = e match {
-      case Lambda(x, _) =>
-        traversal.Continue(e, CountingVisitor(
-          values + (x -> (values.getOrElse(x, 0) + 1)),
-          others
-        ))
-      case DepLambda(x: Kind.Identifier, _) =>
-        traversal.Continue(e, CountingVisitor(
-          values,
-          others + (x -> (others.getOrElse(x, 0) + 1))
-        ))
-      case _ => traversal.Continue(e, this)
+    var values: Map[Identifier, Int],
+    var types: Map[Kind.Identifier, Int]
+  ) extends PureTraversal {
+    override def binding[I <: Identifier]: I => Pure[I] = i => {
+      values = values + (i -> (values.getOrElse(i, 0) + 1))
+      return_(i)
     }
-
-    override def visitType[T <: Type](t: T): traversal.Result[T] = {
-      case class TypeVisitor(
-        others: Map[Kind.Identifier, Int]
-      ) extends traversal.Visitor {
-        override def visitType[U <: Type](t: U): traversal.Result[U] = t match {
-          case DepFunType(x: Kind.Identifier, _) =>
-            traversal.Continue(t, TypeVisitor(
-              others + (x -> (others.getOrElse(x, 0) + 1))
-            ))
-          case _ => traversal.Continue(t, this)
-        }
-      }
-      traversal.types.DepthFirstGlobalResult(t, TypeVisitor(others))
-      traversal.Continue(t, this)
+    override def depBinding[I <: Kind.Identifier]: I => Pure[I] = i => {
+      types = types + (i -> (types.getOrElse(i, 0) + 1))
+      return_(i)
     }
   }
 
   def check(e: Expr): Boolean = {
-    traversal.DepthFirstGlobalResult(e, CountingVisitor(Map(), Map())) match {
-      case traversal.Continue(_, CountingVisitor(values, others)) =>
-        val valuesDup = values.filter({ case (_, n) => n > 1 })
-        val othersDup = others.filter({ case (_, n) => n > 1 })
-        valuesDup.isEmpty && othersDup.isEmpty
-      case _ => throw new Exception("")
-    }
+    val cv = CountingVisitor(Map(), Map())
+    Traverse(e, cv)
+    val valuesDup = cv.values.filter({ case (_, n) => n > 1 })
+    val typesDup = cv.types.filter({ case (_, n) => n > 1 })
+    valuesDup.isEmpty && typesDup.isEmpty
```

Comment:
here it looks like the `var`s are used as global traversal state, which seems incompatible with how they are used in `IsClosedForm`, is there a bug in `IsClosedForm` then?

Reply:
From what I understand `uniqueNames` here checks that binders are unique globally (e.g., in `(\x.x)(\x.x)` binders are not globally unique) while `IsClosedForm` checks that the references in each branch of the AST reference binders up the tree, which is (as you say) local.